File: testsuite/tests/ada/0003-private-types/cpp/main.cpp
#include <iostream>

#include "test.h"

test::Pair t(test::Pair p) {
  p.print();
  // Initialize the pair using the default allocation function
  test::Pair p2;
  p2.print();
  // Verify that a class passed by argument works
  p.add_pairs(test::init_pair(4, 3)).print();

  return p;
}

int main() {
  // Get a Pair from a free function
  test::Pair p = test::init_pair(3, 4);
  std::cout << p.left() << "\n";
  std::cout << p.sum() << "\n";
  std::cout.flush();
  // The value should be copied, and the copy freed by the callee.
  t(p);
  // `p` should still be allocated.
  p.print();
}

File: testsuite/tests/ada/0003-private-types/java/Main.java
import com.adacore.libtest.test.Pair;
import com.adacore.libtest.test.TestPackage;
import com.adacore.libtest.derivation.OtherPair;
import com.adacore.libtest.derivation.DerivationPackage;

public class Main {
    static Pair t(Pair p) {
        p.print();
        Pair p2 = new Pair();
        p2.print();
        p.addPairs(TestPackage.initPair(4, 3)).print();

        return p;
    }

    public static void main(String[] args) {
        Pair p = TestPackage.initPair(3, 4);
        System.out.format("%d\n", p.left());
        System.out.format("%d\n", p.sum());

        t(p);
        p.print();

        OtherPair op = DerivationPackage.initPair(1, 2);
        op.print();
    }
}
